### ISSUE TITLE
New threads being created unexpectedly?

### ISSUE THREAD

**Issue by smeriwether** [celluloid/celluloid-supervision#27]
Hello, I am pretty new to Celluloid (and celluloid-supervision) so to get started I created a simple example to try and get a feel for how things work.

I will post my code below but here is a rundown of my example. 
- 1 supervision container (manages 2 actors)
- 2 actors
- Every x seconds, one actor sends a message to the other actor to restart itself

```
require_relative 'lib/supervisor_group'

def start_supervising
  loop do
    group = SupervisorGroup.run
    sleep 5 while group.alive?
    puts "!!! Celluloid::Supervision::Container #{self} crashed. Restarting..."
  end
end

start_supervising
```

```
require 'celluloid/current'

require_relative 'actors/actor_one'
require_relative 'actors/actor_two'

class SupervisorGroup
  def self.run
    create
  end

  def self.create
    container = Celluloid::Supervision::Container.run!
    container.add(configuration(ActorOne, :account1_actor_one, container, 'account1'))
    container.add(configuration(ActorTwo, :account1_actor_two, container, 'account1'))
    container
  end

  def self.configuration(actor, actor_as, *args)
    {
      type: actor,
      as: actor_as,
      args: { args: args }
    }
  end
end
```

```
require 'celluloid/current'

class ActorOne
  include Celluloid
  finalizer :finalizer

  def initialize(args)
    @account = args[1][1]
    @container = args[1][0]
    async.speak
  end

  def speak
    every(60) do
      puts "I am talking in 1 from #{@account}"
    end
  end

  def testing
    puts "I am testing inside of 1 from #{@account}"
    current_actor = Actor.current
    @container.restart_actor(current_actor, 'New account data')
  end

  def finalizer
    puts "I am ending from 1 for #{@account}"
  end
end
```

```
require 'celluloid/current'

class ActorTwo
  include Celluloid

  def initialize(args)
    @account = args[1][1]
    @container = args[1][0]
    async.speak
  end

  def speak
    every(5) do
      @container.actors.first.testing
    end
  end
end
```

This code works and does what I expect from a users perspective. 

Upon inspecting the ruby process using Activity Monitor (on OSX) I notice that every time the I call `@container.restart_actor` I am seeing new OS threads and an increase in memory usage. 

So I added the `finalizer` method to do a bit more digging and found that once I quit the process, that finalizer method is being called for every `@container.restart_actor` call. (So if the testing method on ActorOne is called 10 times then the its finalizer method also gets called 10 times)

This behavior isn't really what I was expected, since the method name is `restart_actor`, my assumption was that it would kill and and create a new actor at the time it was called. My intention is to have a very long running process that follows this pattern every so often so I fear the memory usage will get out-of-control relatively quickly. 
- I am using ruby-2.2.2
- celluloid 0.17.2
- celluloid-supervision 0.20.5

If there are some things here I am doing wrong I would love to know. There wasn't really much documentation on this so I did what made since after spending a bit looking at the source. I would definitely be up to adding this type of example to the current examples if it looks like this is how it should be written.

**Comment by smeriwether** [comment]
A side note, I was able to refactor why use case to call `restart_actor` much less than I was initially intending (and maybe with a bit more thought I could remove the use of `restart_actor` completely) but I wanted to still post this here to see if I have some sort of misunderstanding.

**Comment by digitalextremist** [comment]
Would you mind posting some explicit statistics to go along with your account of the behavior?

The `finalizer` ought to fire on `restart_actor` because the actors are not killed but are terminated, which is a voluntary action of the actor itself -- it's just a message put into the mailbox that says "kill yourself" ... which invokes the `finalizer` ... but the accrual of threads by this is not right. It's almost as if your actor(s) are hitting an error you aren't seeing -- but even that ought not accrue threads.

I'll look at this while you post statistics... and if you don't mind, can you also try `rbx` and/or `jruby` to be sure this isn't an `mri` behavior issue? I appreciate you going through code to work to understand the issue in the absence of documentation for this specific gem.

**Comment by smeriwether** [comment]
Thanks for responding! Sure thing, I will gather the stats and try `rbx` and `jruby` like you suggested. 

One thing I wanted to note because I don't think I was very clear on what I was seeing:

The finalizer method wasn't being called on `restart_actor`. What I saw (and I can verify while I am getting the things mentioned above) was the finalizer only being called when I manually `CTRL-C` (quit) the process, then I would see the 10 finalizer print messages. That may be expected
behavior but it seems weird which is why I included that bit in my message.

**Comment by digitalextremist** [comment]
Oh, that behavior is an output buffer issue. You may see different behavior on `rbx` and `jruby` with that. The output buffer behaves differently per ruby engine.

**Comment by smeriwether** [comment]
I have some stats from `ruby` and `jruby`. I haven't had time to try it out with `rbx` yet but I will soon. What I did was simply run the exact code above and screenshot OS Xs Activity Monitor as I started the program and about 30 minutes later before I ended it. (I ran `jruby` ~25 min and `ruby` ~30 min). Here are the stats:

```
Ruby:
* Memory (Start): 18.8 MB
* Memory (End): 50.1 MB
* Threads (Start): 9
* Threads (End): 422
```

```
JRuby:
* Memory (Start): 184.1 MB
* Memory (End): 263.0 MB
* Threads (Start): 35
* Threads (End): 213
```

The thread count roughly ticked +1 every 5 seconds (every `restart_actor`). The memory usage for engines wasn't completely linear. For the `ruby` process I saw 5 MB spikes or dips sporadically and for `jruby` it was about 10 MB but overall it was trending up.

<img width="814" alt="ruby-start" src="https://cloud.githubusercontent.com/assets/7351237/10798356/3746b0de-7d75-11e5-87b8-760959317e43.png">
<img width="804" alt="ruby-end" src="https://cloud.githubusercontent.com/assets/7351237/10798367/3ea6fb5e-7d75-11e5-83f1-5e62af16a73b.png">
<img width="801" alt="jruby-start" src="https://cloud.githubusercontent.com/assets/7351237/10798374/45243460-7d75-11e5-818f-5ff5d4982057.png">
<img width="799" alt="jruby-stop" src="https://cloud.githubusercontent.com/assets/7351237/10798377/48cd52c2-7d75-11e5-9462-839492236bae.png">


If you wanted more detailed stats I certainly wouldn't mind providing them. Although I may need some guidance on what to do to get those to you. Again, thanks for looking into this!
